trie/database.go

```diff
@@ -768,9 +723,41 @@ func (c *cleaner) Delete(key []byte) error {
 	panic("not implemented")
 }
 
+// Size returns the current storage size of the memory cache in front of the
+// persistent database layer.
+func (db *hashDatabase) Size() common.StorageSize {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	// db.dirtiesSize only contains the useful data in the cache, but when reporting
+	// the total memory consumption, the maintenance metadata is also needed to be
+	// counted.
+	var metadataSize = common.StorageSize((len(db.dirties) - 1) * cachedNodeSize)
+	var metarootRefs = common.StorageSize(len(db.dirties[common.Hash{}].children) * (common.HashLength + 2))
+	return db.dirtiesSize + db.childrenSize + metadataSize - metarootRefs
+}
+
+// IsEmpty returns an indicator if the node database is empty.
+// Hash database is only regarded as empty if *non-empty* genesis
+// state is missing.
+func (db *hashDatabase) IsEmpty() bool {
+	hash := rawdb.ReadCanonicalHash(db.diskdb, 0)
+	if hash == (common.Hash{}) {
+		return true // genesis block is not existent
+	}
+	block := rawdb.ReadBlock(db.diskdb, hash, 0)
+	if block == nil {
+		return true // genesis block is not existent
+	}
+	if convertEmpty(block.Root()) == types.EmptyRootHash {
+		return false // genesis block is empty
+	}
```

Comment:
This method is called by trie/database_wrapper.go, `IsEmpty`, which is unused for now, so I'm not sure how this is intended to be used... BUT: in a pow network, it's possible to not have a genesis alloc, and thus emptyRootHash as state root. So not sure this check should be here... ?

Reply:
Also, I think you also flipped it. It should return `true` to match the comment.

Reply:
Ah, 
```
if header.Root != types.EmptyRootHash && !triedb.Initialized(header.Root) {
```
When used like that, only with non-empty state, it's fine then.

Reply:
Ah wait, your `return false` makes it fine either way. I thought it was a mistake, but now I get it.